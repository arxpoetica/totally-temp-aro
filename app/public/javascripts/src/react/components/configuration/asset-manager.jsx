import React, { Component } from 'react'
import reduxStore from '../../../redux-store'
import AroHttp from '../../common/aro-http'
import wrapComponentWithProvider from '../../common/provider-wrapped-component'
import ConfigurationActions from './configuration-actions'

export class AssetManager extends Component {
  constructor (props) {
    super(props)
    this.fileInput = React.createRef()
    this.state = {
      assetKeys: [],
      isValidFileSelected: false
    }
    AroHttp.get('/ui_assets/list/assetKeys?limit=500')
      .then(result => {
        this.setState({
          assetKeys: result.data
        })
      })
      .catch(err => console.error(err))
  }

  render () {
    return <div>
      <div style={{ maxHeight: '300px', overflowY: 'auto' }}>
        <table className='table table-sm table-striped'>
          <thead className='thead thead-dark'>
            <tr>
              <th>Asset key</th>
              <th style={{ textAlign: 'center' }}>Image</th>
            </tr>
          </thead>
          <tbody>
            {this.state.assetKeys.map(assetKey =>
              <tr key={assetKey}>
                <td style={{ verticalAlign: 'middle' }}>{assetKey}</td>
                <td style={{ textAlign: 'center' }}><img src={`/ui_assets/${assetKey}`} /></td>
              </tr>
            )}
          </tbody>
        </table>
      </div>
      <hr />
      <h4>Upload a new file:</h4>
      <input type='file' ref={this.fileInput} onChange={event => this.setState({ isValidFileSelected: Boolean(this.fileInput.current) })} />
      <button className='btn btn-primary' disabled={!this.state.isValidFileSelected} onClick={event => this.uploadFile()}>
        Upload
      </button>
    </div>
  }

  uploadFile () {
    const file = this.fileInput.current.files[0]
    const assetKey = file.name
    this.props.uploadAssetToServer(assetKey, file)
  }
}

AssetManager.propTypes = {
}

const mapStateToProps = (state) => ({
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  uploadAssetToServer: (assetKey, file) => dispatch(ConfigurationActions.uploadAssetToServer(assetKey, file))
})

const AssetManagerComponent = wrapComponentWithProvider(reduxStore, AssetManager, mapStateToProps, mapDispatchToProps)
export default AssetManagerComponent
