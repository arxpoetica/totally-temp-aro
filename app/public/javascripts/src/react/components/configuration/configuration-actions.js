/* globals FormData */
import Actions from '../../common/actions'
import AroHttp from '../../common/aro-http'

function loadConfigurationFromServer () {
  return dispatch => {
    AroHttp.get('/ui_settings')
      .then(result => dispatch({
        type: Actions.CONFIGURATION_SET_CONFIGURATION,
        payload: result.data
      }))
      .catch(err => console.error(err))
  }
}

function saveConfigurationToServerAndReload (type, configuration) {
  return dispatch => {
    AroHttp.post(`/ui_settings/save/${type}`, { configuration: configuration })
      .then(result => dispatch(loadConfigurationFromServer))
      .catch(err => console.error(err))
  }
}

function getAssetKeys (offset, limit) {
  return dispatch => {
    AroHttp.get(`/ui_assets/list/assetKeys?offset=${offset}&limit=${limit}`)
      .then(result => dispatch({
        type: Actions.CONFIGURATION_SET_ASSET_KEYS,
        payload: result.data
      }))
      .catch(err => console.error(err))
  }
}

function uploadAssetToServer (assetKey, file) {
  return dispatch => {
    var formData = new FormData()
    formData.append('file', file)
    AroHttp.postRaw(`/ui_assets/${assetKey}`, formData) // Important to send empty headers so file upload works
      .then(() => dispatch(getAssetKeys(0, 500)))
      .catch(err => console.error(err))
  }
}

function getReportsMetadata () {
  return dispatch => {
    AroHttp.get('/service/v2/installed/report/meta-data')
      .then(result => dispatch({
        type: Actions.CONFIGURATION_SET_REPORTS_METADATA,
        payload: result.data
      }))
      .catch(err => console.error(err))
  }
}

function startEditingReport (reportId) {
  return {
    type: Actions.CONFIGURATION_SET_EDITING_REPORT_ID,
    payload: reportId
  }
}

function populateEditingReportDefinition (reportId) {
  return dispatch => {
    AroHttp.get(`/service/v2/report-module/${reportId}`)
      .then(result => dispatch({
        type: Actions.CONFIGURATION_SET_EDITING_REPORT_DEFINITION,
        payload: result.data
      }))
      .catch(err => console.error(err))
  }
}

function clearEditingReportDefinition () {
  return {
    type: Actions.CONFIGURATION_SET_EDITING_REPORT_DEFINITION,
    payload: null
  }
}

function saveEditingReportPrimaryDefinition (primaryDefinition) {
  return {
    type: Actions.CONFIGURATION_SET_EDITING_REPORT_PRIMARY_DEFINITION,
    payload: primaryDefinition
  }
}

function saveEditingReportSubDefinition (subDefinition, subDefinitionIndex) {
  return {
    type: Actions.CONFIGURATION_SET_EDITING_REPORT_SUBDEFINITION,
    payload: {
      subDefinition: subDefinition,
      subDefinitionIndex: subDefinitionIndex
    }
  }
}

function saveCurrentReportToServer () {
  return (dispatch, getState) => {
    // We have to do a getState() because there may be state changes that have not yet been updated in the calling component
    const reportDefinition = getState().configuration.reports.reportBeingEdited
    return AroHttp.put(`/service/v2/report-module/${reportDefinition.id}`, reportDefinition)
      .then(() => dispatch({
        type: Actions.CONFIGURATION_CLEAR_EDITING_REPORT
      }))
      .catch(err => console.error(err))
  }
}

function deleteReport (reportId) {
  return dispatch => {
    AroHttp.delete(`/service/v2/report-module/${reportId}`)
      .then(() => dispatch(getReportsMetadata()))
      .catch(err => console.error(err))
  }
}

function validateReport (planId) {
  return (dispatch, getState) => {
    // We have to do a getState() because there may be state changes that have not yet been updated in the calling component
    const reportDefinition = getState().configuration.reports.reportBeingEdited
    return AroHttp.post(`/service/v2/report-module-validate/${planId}?sampleSize=10`, reportDefinition)
      .then(result => dispatch({
        type: Actions.CONFIGURATION_SET_REPORT_VALIDATION,
        payload: result.data
      }))
      .catch(err => {
        console.error(err)
        dispatch({
          type: Actions.CONFIGURATION_SET_REPORT_VALIDATION,
          payload: err
        })
      })
  }
}

export default {
  loadConfigurationFromServer: loadConfigurationFromServer,
  saveConfigurationToServerAndReload: saveConfigurationToServerAndReload,
  getAssetKeys: getAssetKeys,
  uploadAssetToServer: uploadAssetToServer,
  getReportsMetadata: getReportsMetadata,
  startEditingReport: startEditingReport,
  populateEditingReportDefinition: populateEditingReportDefinition,
  clearEditingReportDefinition: clearEditingReportDefinition,
  saveEditingReportPrimaryDefinition: saveEditingReportPrimaryDefinition,
  saveEditingReportSubDefinition: saveEditingReportSubDefinition,
  saveCurrentReportToServer: saveCurrentReportToServer,
  deleteReport: deleteReport,
  validateReport: validateReport
}
