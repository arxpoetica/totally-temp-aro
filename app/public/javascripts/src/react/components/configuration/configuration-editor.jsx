import React, { Component } from 'react'
import { PropTypes } from 'prop-types'
import reduxStore from '../../../redux-store'
import wrapComponentWithProvider from '../../common/provider-wrapped-component'
import ConfigurationActions from '../configuration/configuration-actions'
import SettingsEditor from './settings-editor.jsx'

export class ConfigurationEditor extends Component {
  constructor (props) {
    super(props)
    // Form component state does not need to go into Redux
    this.state = {
      showInitialWarning: true
    }
  }

  render () {
    return <div>
      { this.state.showInitialWarning
        ? this.renderWarningMessage()
        : <SettingsEditor id='compSettingsEditor' />
      }
    </div>
  }

  renderWarningMessage () {
    return <div id='divConfigurationWarning'>
      <div className='alert alert-danger'>
        <strong>Careful:</strong> These settings are intended to be changed by system administrators. Incorrect settings
        can cause the entire platform to stop working <strong>for all users</strong>.
      </div>
      <button id='btnAcceptConfigurationWarning' className='btn btn-block btn-danger'
        onClick={event => this.setState({ showInitialWarning: false })}>
        I understand, take me to the configuration editor
      </button>
    </div>
  }
}

ConfigurationEditor.propTypes = {
  configuration: PropTypes.object
}

const mapStateToProps = (state) => ({
  initialConfiguration: state.configuration
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  saveConfigurationToServerAndReload: (type, configuration) => dispatch(ConfigurationActions.saveConfigurationToServerAndReload(type, configuration))
})

const ConfigurationEditorComponent = wrapComponentWithProvider(reduxStore, ConfigurationEditor, mapStateToProps, mapDispatchToProps)
export default ConfigurationEditorComponent
