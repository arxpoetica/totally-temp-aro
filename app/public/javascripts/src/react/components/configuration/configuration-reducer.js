import Actions from '../../common/actions'

const defaultState = {
  items: [],
  assetKeys: []
}

function setConfiguration (state, configuration) {
  return { ...state,
    items: configuration
  }
}

function setAssetKeys (state, assetKeys) {
  return { ...state,
    assetKeys: assetKeys
  }
}

function configurationReducer (state = defaultState, action) {
  switch (action.type) {
    case Actions.CONFIGURATION_SET_CONFIGURATION:
      return setConfiguration(state, action.payload)

    case Actions.CONFIGURATION_SET_ASSET_KEYS:
      return setAssetKeys(state, action.payload)

    default:
      return state
  }
}

export default configurationReducer
