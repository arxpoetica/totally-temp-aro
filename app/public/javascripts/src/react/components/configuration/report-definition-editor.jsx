import React, { Component } from 'react'
import { Field, reduxForm } from 'redux-form'
import Constants from '../../common/constants'
import './report-definition-editor.css'

const renderField = ({
  input,
  label,
  type,
  meta: { touched, error, warning }
}) => (
  <div>
    <input {...input} placeholder={label} type={type} className='form-control' />
    {error && <div className='text-danger' style={{ fontSize: '10px', fontWeight: 'bold' }}>{error}</div>}
  </div>
)

export class ReportDefinitionEditor extends Component {
  render () {
    return <form className='d-flex flex-column report-definition-editor'
      style={{ height: '100%' }}
      onSubmit={event => event.preventDefault()}>
      <div className='form-row flex-grow-0'>
        <div className='col'>
          <label>Name</label>
          <Field name='name' className='form-control' type='text' validate={[this.validateName]} component={renderField} />
        </div>
        <div className='col'>
          <label>Display Name</label>
          <Field name='displayName' className='form-control' component='input' type='text' />
        </div>
        <div className='col'>
          <label>Query Type</label>
          <Field name='queryType' className='form-control' component='select' type='text'>
            <option value='SQL_REPORT'>SQL Report</option>
            <option value='SQL_SCRIPT'>SQL Script</option>
            <option value='KML_REPORT'>KML Report</option>
          </Field>
        </div>
      </div>
      <div className='form-row flex-grow-1' style={{ paddingTop: '10px', paddingBottom: '10px' }}>
        <div className='col' style={{ height: '100%' }}>
          <Field name='query' className='form-control' component='textarea' type='text' style={{ height: '100%', fontFamily: 'Courier New', fontSize: '12px' }} />
        </div>
      </div>
    </form>
  }

  validateName (value) {
    value = value || ''
    return (value.indexOf(' ') >= 0) ? 'Name cannot have spaces' : undefined
  }
}

let ReportDefinitionEditorForm = reduxForm({
  form: Constants.REPORT_DEFINITION_EDITOR_FORM
})(ReportDefinitionEditor)

export default ReportDefinitionEditorForm
