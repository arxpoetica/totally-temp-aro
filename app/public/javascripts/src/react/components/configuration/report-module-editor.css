.report-module-editor label {
  font-weight: 700;
}
