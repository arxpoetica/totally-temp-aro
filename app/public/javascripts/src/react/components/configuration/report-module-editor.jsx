import React, { Component } from 'react'
import { connect } from 'react-redux'
import { PropTypes } from 'prop-types'
import ConfigurationActions from './configuration-actions'
import ReportDefinitionEditor from './report-definition-editor.jsx'

export class ReportModuleEditor extends Component {
  constructor (props) {
    super(props)
    this.props.populateEditingReportDefinition(this.props.reportId)
    this.state = {
      isEditingPrimary: true,
      subDefinitionEditingIndex: -1
    }
  }

  getDefinitionBeingEdited () {
    if (!this.props.moduleDefinition) {
      return null
    } else {
      return this.state.isEditingPrimary
        ? this.props.moduleDefinition.definition
        : this.props.moduleDefinition.subDefinitions[this.state.subDefinitionEditingIndex]
    }
  }

  render () {
    return <div className='container' style={{ height: '100%' }}>
      <div className='row' style={{ height: '100%' }}>
        <div className='col-md-2'>
          <ul className='nav nav-pills'>
            <li className='nav-item' key='-1'>
              <a className='nav-link active' onClick={() => this.setState({ isEditingPrimary: true, subDefinitionEditingIndex: -1 })}>Primary Definition</a>
            </li>
            {
              this.props.moduleDefinition
                ? this.props.moduleDefinition.subDefinitions.map((subDefinition, index) => (
                  <li className='nav-item' key={subDefinition.id}>
                    <a className='nav-link' onClick={() => this.setState({ isEditingPrimary: false, subDefinitionEditingIndex: index })}>
                      Subdefinition {index}
                    </a>
                  </li>
                ))
                : null
            }
          </ul>
        </div>
        <div className='col-md-10 d-flex flex-column' style={{ height: '100%' }}>
          { this.getDefinitionBeingEdited()
            ? <div className='flex-grow-1'><ReportDefinitionEditor initialValues={this.getDefinitionBeingEdited()} enableReinitialize /></div>
            : null
          }
          {/* Show an alert if required */}
          {
            <div className='form-row flex-grow-0'>
              <div className='alert alert-primary' role='alert'>
                This is a primary alertâ€”check it out!
              </div>
            </div>
          }
          <div className='form-row flex-grow-0'>
            <div className='float-right'>
              <button className='btn btn-light'>Test</button>
              <button className='btn btn-primary'>Save Definition</button>
            </div>
          </div>
        </div>
      </div>
    </div>
  }

  submit (values) {
    // print the form values to the console
    console.log(values)
  }
}

ReportModuleEditor.propTypes = {
  reportId: PropTypes.number,
  reportDefinition: PropTypes.object
}

const mapStateToProps = (state) => ({
  reportId: state.configuration.reports.reportBeingEdited && state.configuration.reports.reportBeingEdited.id,
  moduleDefinition: (state.configuration.reports.reportBeingEdited &&
    state.configuration.reports.reportBeingEdited.definition &&
    state.configuration.reports.reportBeingEdited.definition.moduleDefinition)
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  populateEditingReportDefinition: reportId => dispatch(ConfigurationActions.populateEditingReportDefinition(reportId))
})

const ReportModuleEditorComponent = connect(mapStateToProps, mapDispatchToProps)(ReportModuleEditor)
export default ReportModuleEditorComponent
