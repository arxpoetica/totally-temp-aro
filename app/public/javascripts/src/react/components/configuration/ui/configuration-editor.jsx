import React, { Component } from 'react'
import reduxStore from '../../../../redux-store'
import wrapComponentWithProvider from '../../../common/provider-wrapped-component'
import UiActions from './ui-actions'
import SettingsEditor from './settings-editor.jsx'
import AssetManager from './asset-manager.jsx'
import StylesheetsManager from './stylesheets-manager.jsx'
import EtlTemplateManager from './etl-template-manager.jsx'

export class ConfigurationEditor extends Component {
  constructor (props) {
    super(props)
    this.navs = Object.freeze({
      SETTINGS: 'Settings',
      ASSET_MANAGER: 'Asset Manager',
      STYLESHEETS: 'Stylesheets',
      ETL_TEMPLATE_MANAGER: 'Templates'
    })
    this.state = {
      showInitialWarning: true,
      selectedNav: this.navs.SETTINGS
    }
  }
  navSelection () {
    let val = this.state.selectedNav

    if (val === 'Settings') {
      return (
        <SettingsEditor id='compSettingsEditor' />
      )
    } else if (val === 'Asset Manager') {
      return (
        <AssetManager />
      )
    } else if (val === 'Stylesheets') {
      return (
        <StylesheetsManager />
      )
    } else {
      return (
        <EtlTemplateManager />
      )
    }
  }
  render () {
    return <div>
      { this.state.showInitialWarning
        ? this.renderWarningMessage()
        : this.renderControls()
      }
    </div>
  }

  renderWarningMessage () {
    return <div id='divConfigurationWarning'>
      <div className='alert alert-danger'>
        <strong>Careful:</strong> These settings are intended to be changed by system administrators. Incorrect settings
        can cause the entire platform to stop working <strong>for all users</strong>.
      </div>
      <button id='btnAcceptConfigurationWarning' className='btn btn-block btn-danger'
        onClick={event => this.setState({ showInitialWarning: false })}>
        I understand, take me to the configuration editor
      </button>
    </div>
  }

  renderControls () {
    return <div>
      <ul className='nav nav-tabs' style={{ marginBottom: '10px' }}>
        <li className='nav-item' onClick={() => this.setState({ selectedNav: this.navs.SETTINGS })}>
          <a className={`nav-link ${this.state.selectedNav === this.navs.SETTINGS ? 'active' : ''}`} href='#'>
            {this.navs.SETTINGS}
          </a>
        </li>
        <li className='nav-item' onClick={() => this.setState({ selectedNav: this.navs.ASSET_MANAGER })}>
          <a className={`nav-link ${this.state.selectedNav === this.navs.ASSET_MANAGER ? 'active' : ''}`} href='#'>
            {this.navs.ASSET_MANAGER}
          </a>
        </li>
        <li className='nav-item' onClick={() => this.setState({ selectedNav: this.navs.STYLESHEETS })}>
          <a className={`nav-link ${this.state.selectedNav === this.navs.STYLESHEETS ? 'active' : ''}`} href='#'>
            {this.navs.STYLESHEETS}
          </a>
        </li>
        <li className='nav-item' onClick={() => this.setState({ selectedNav: this.navs.ETL_TEMPLATES })}>
          <a className={`nav-link ${this.state.selectedNav === this.navs.ETL_TEMPLATES ? 'active' : ''}`} href='#'>
            {this.navs.ETL_TEMPLATE_MANAGER}
          </a>
        </li>
      </ul>
      {
        this.navSelection()
        // (this.state.selectedNav === this.navs.SETTINGS)
        //   ? <SettingsEditor id='compSettingsEditor' />
        //   : <AssetManager />
      }
    </div>
  }
}

ConfigurationEditor.propTypes = {}

const mapStateToProps = (state) => ({})

const mapDispatchToProps = (dispatch, ownProps) => ({
  saveConfigurationToServerAndReload: (type, configuration) => dispatch(UiActions.saveConfigurationToServerAndReload(type, configuration))
})

const ConfigurationEditorComponent = wrapComponentWithProvider(reduxStore, ConfigurationEditor, mapStateToProps, mapDispatchToProps)
export default ConfigurationEditorComponent
