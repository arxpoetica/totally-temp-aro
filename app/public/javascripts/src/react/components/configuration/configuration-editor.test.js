/* global test expect */
import React from 'react'
import { shallow } from 'enzyme'
import { ConfigurationEditor } from '../configuration/configuration-editor'

const sampleConfiguration = {
  locationCategories: {
    medium_business: {
      label: 'Medium businesses'
    }
  }
}

// -----------------------------------------------------------------------------
test('Component render - default and after clicking button', () => {
  const component = shallow(
    <ConfigurationEditor initialConfiguration={sampleConfiguration} />
  )
  // Initial state - show warning, hide form
  expect(component).toMatchSnapshot()
  expect(component.find('#divConfigurationWarning').exists()).toBeTruthy()
  expect(component.find('#divConfigurationForm').exists()).toBeFalsy()
  component.find('#btnAcceptConfigurationWarning').simulate('click')

  // New state - hide warning, show form
  expect(component.find('#divConfigurationWarning').exists()).toBeFalsy()
  expect(component.find('#divConfigurationForm').exists()).toBeTruthy()
  expect(component).toMatchSnapshot()
})
