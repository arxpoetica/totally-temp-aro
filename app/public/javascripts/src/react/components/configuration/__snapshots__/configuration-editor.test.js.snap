// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component render - default and after clicking button 1`] = `
<div>
  <div
    id="divConfigurationWarning"
  >
    <div
      className="alert alert-danger"
    >
      <strong>
        Careful:
      </strong>
       These settings are intended to be changed by system administrators. Incorrect settings can cause the entire platform to stop working 
      <strong>
        for all users
      </strong>
      .
    </div>
    <button
      className="btn btn-block btn-danger"
      id="btnAcceptConfigurationWarning"
      onClick={[Function]}
    >
      I understand, take me to the configuration editor
    </button>
  </div>
</div>
`;

exports[`Component render - default and after clicking button 2`] = `
<div>
  <div
    id="divConfigurationForm"
  >
    <form>
      <div
        className="form-group row"
      >
        <label
          className="col-sm-4 col-form-label"
        >
          Configuration Type
        </label>
        <div
          className="col-sm-8"
        >
          <select
            className="form-control"
            id="selectConfigurationType"
            onChange={[Function]}
            value="locationCategories"
          >
            <option
              key="locationCategories"
              value="locationCategories"
            >
              locationCategories
            </option>
          </select>
        </div>
      </div>
      <div
        className="form-group row"
      >
        <label
          className="col-sm-4 col-form-label"
        >
          Value
        </label>
        <div
          className="col-sm-8"
        >
          <textarea
            className="form-control"
            onChange={[Function]}
            style={
              Object {
                "height": "350px",
              }
            }
            value="{
  \\"medium_business\\": {
    \\"label\\": \\"Medium businesses\\"
  }
}"
          />
        </div>
      </div>
    </form>
    <button
      className="btn btn-primary float-right"
      onClick={[Function]}
    >
      <i
        className="fa fa-save"
      />
      Save settings
    </button>
  </div>
</div>
`;
